mybuild_docs/sphinx/man/2.10.0/fpgaport.8
.\" Man page generated from reStructuredText.
.
.TH "FPGAPORT" "8" "Nov 22, 2023" "2.10.0" "OPAE"
.SH NAME
fpgaport \- Enable/Disable virtualization
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
fpgaport [\-h] [\-N NUMVFS] [\-X] [\-\-debug] {assign,release} device [port]
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
The \fBfpgaport\fP enables and disables virtualization. It assigns
and releases control of the port to the virtual function (VF). By default, the driver
assigns the port to the physical function (PF) in the non\-virtualization use case.
.SH POSITIONAL ARGUMENTS
.sp
\fB{assign, release}\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Action to perform.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBdevice\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
The FPGA device being targeted with this action.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBport\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
The number of the port.
.ft P
.fi
.UNINDENT
.UNINDENT
.SH OPTIONAL ARGUMENTS
.sp
\fB\-N NUMVFS, \-\-numvfs NUMVFS\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Create NUMVFS virtual functions. The typical value is 1.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-X, \-\-destroy\-vfs\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Destroy all virtual functions prior to assigning.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-\-debug\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Display additional log information.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fB\-h, \-\-help\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Print usage information.
.ft P
.fi
.UNINDENT
.UNINDENT
.SH EXAMPLE
.sp
\fBfpgaport release /dev/dfl\-fme.0 0\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Release port 0 from physical function control.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBfpgaport assign /dev/dfl\-fme.0 0\fP
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Assign port 0 to physical function control.
.ft P
.fi
.UNINDENT
.UNINDENT
.SH REVISION HISTORY
.sp
| Document Version |  Intel Acceleration Stack Version  | Changes  |
| —————\- |————————————|———\-|
| 2018.05.21 | 1.1 Beta. (Supported with Intel Quartus Prime Pro Edition 17.1.) | No changes from previous release.  |
.SH AUTHOR
Intel DCG FPT SW
.SH COPYRIGHT
2017 Intel Corporation
.\" Generated by docutils manpage writer.
.


python/opae.admin/opae/admin/tools/fpgaport.py
#!/usr/bin/env python3
# Copyright(c) 2019, Intel Corporation
#
# Redistribution  and  use  in source  and  binary  forms,  with  or  without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of  source code  must retain the  above copyright notice,
#   this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright notice,
#   this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
# * Neither the name  of Intel Corporation  nor the names of its contributors
#   may be used to  endorse or promote  products derived  from this  software
#   without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
# IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
# LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
# CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
# SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
# INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
# CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
from __future__ import absolute_import
import argparse
import logging
import os
import sys

from opae.admin.fpga import fpga
from opae.admin.sysfs import pci_node


LOG = logging.getLogger('fpgaport')


def fpga_filter(inp):
    filt = {'pci_node.supports_sriov': True}
    if inp == 'all':
        return filt
    if inp.startswith('/dev') and os.path.exists(inp):
        filt.update({'fme.devpath': inp})
    else:
        pci_info = pci_node.parse_address(inp)
        if pci_info:
            pci_address = pci_info['pci_address']
            filt.update({'pci_node.pci_address': pci_address})
        else:
            raise TypeError('{} is not valid device pattern')
    return filt


def set_numvfs(device, value):
    try:
        device.pci_node.sriov_numvfs = value
    except ValueError as err:
        LOG.warning('error setting num_sriov: "%s"', err)


def assign(args, device):
    destroy_vfs = args.numvfs == 0 or args.destroy_vfs
    if not destroy_vfs and device.pci_node.sriov_numvfs:
        LOG.warning('Device (%s) has VFs, skipping...', device.pci_node)
        return
    numvfs = device.pci_node.sriov_numvfs
    set_numvfs(device, 0)
    try:
        device.fme.assign_port(args.port)
    except (OSError, IOError) as err:
        LOG.warning('assiging port %d for device %s: %s', args.port,
                    device.pci_node, err)
        set_numvfs(device, numvfs)


def release(args, device):
    if args.destroy_vfs:
        LOG.warning('--destroy-vfs only applicable with "assign"')
    try:
        device.fme.release_port(args.port)
    except (OSError, IOError) as err:
        LOG.warning('releasing port %d for device %s: %s', args.port,
                    device.pci_node, err)
    else:
        if args.numvfs:
            set_numvfs(device, args.numvfs)


def main():
    actions = {'assign': assign,
               'release': release}
    # parse command line arguments
    parser = argparse.ArgumentParser()
    parser.add_argument('action', action='store',
                        choices=sorted(actions.keys()),
                        help='Action to perform - {}'.format(
                            list(actions.keys())))
    parser.add_argument('device',
                        help=('The FPGA (FME) device. '
                              'Can be identified by dev path (/dev/dfl-fme.0) '
                              'or by PCIe address: '
                              '([<segment>:]<bus>:<device>.<function>'))
    parser.add_argument('port', type=int, default=0, nargs='?',
                        help='The port number.')
    parser.add_argument('-N', '--numvfs', type=int,
                        help='Create VFs from the argument given.')
    parser.add_argument('-X', '--destroy-vfs', action='store_true',
                        default=False,
                        help='Destroy all VFs prior to assigning.')
    parser.add_argument('--debug', action='store_true', default=False,
                        help='Set logging to DEBUG if verbose selected')

    args = parser.parse_args()

    level = logging.DEBUG if args.debug else logging.INFO
    logging.basicConfig(level=level,
                        format=('[%(asctime)-15s] [%(levelname)-8s] '
                                '%(message)s'))
    devices = fpga.enum([fpga_filter(args.device)])

    if not devices:
        LOG.error('Could not find device using pattern: "%s"', args.device)
        sys.exit(os.EX_USAGE)

    for device in devices:
        actions[args.action](args, device)


if __name__ == "__main__":
    main()
